docs/DesignDocs/openenclave_modules.md

```diff
@@ -0,0 +1,105 @@
+# OpenEnclave Modules
+
+The Open Enclave SDK provides a core set of functionality for build enclave
+applications. Some of these features&2014;such as memory management and ecall/ocall
+marshalling&2014;are required to create an enclave. Other features such as
+libc functions, host filesystem, and network sockets can be optionally included
+inside the enclave by the enclave developer. These optional features are known
+as **Open Enclave Modules** and are static libraries that can be optionally
+linked into the enclave binary.
+
+## Currently supported modules
+
+Currently, Open Enclave only supports modules that are part of the OE SDK source tree.
+Today there are 5 optionally includable modules:
+
+* hostfs
+* libc
+* hostepoll
+* hostsock
+* hostresolver
+
+## Module initialization
+
+OE modules can each have an initialization function which will be called on
```

Comment:
Just a random thought: would it be possible to support pass-by-value for the initialization functions? I imagine this would complicate the design (i.e., difficult to generalize). However, this may be needed for some use cases (e.g., larger number of functions with dependencies).

Reply:
Do you mean for specifying constructor order? Using the builtin compiler attribute mechanism I think the answer is no, although I may be able to come up with something with a little preprocessor magic